File: Dynamic Programming/Champagne Tower/ChampagneTower.java
// DP. Split overflow between ith glass and i + 1th glass in next row. return dp[query_row][query_glass]
public class ChampagneTower {
	public double champagneTower(int poured, int query_row, int query_glass) {
		double[][] dp = new double[query_row + 2][query_row + 2];
        dp[0][0] = poured;
        for (int i = 0; i <= query_row; i++) {
            for (int j = 0; j <= i; j++) {
                double overflow = dp[i][j] - 1;
                if (overflow > 0) {
                    dp[i + 1][j] += overflow / 2.0;
                    dp[i + 1][j + 1] += overflow / 2.0;
                }
            }
        }
        return Math.min(dp[query_row][query_glass], 1);
	}
}

File: Dynamic Programming/Champagne Tower/champagne_tower.cpp
class Solutio {
public:
    // bottom-up dp: divide overflow among (curr_row + 1, curr_glass) and (curr_row + 1, curr_glass + 1)
    double champagneTower(int poured, int query_row, int query_glass) {
        vector<vector<double>> dp(query_row + 2, vector<double>(query_row + 2, 0));
        dp[0][0] = poured;
        double overflow;
        for (int i = 0; i <= query_row; ++i) {
            for (int j = 0; j <= i; ++j) {
                overflow = max(dp[i][j] - 1, 0.0);
                dp[i + 1][j] += overflow / 2.0;
                dp[i + 1][j + 1] += overflow / 2.0;
            }
        }

        return min(1.0, dp[query_row][query_glass]);
    }
};
